File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/instance/termination/InstanceTerminationState.java
package com.sequenceiq.cloudbreak.core.flow2.stack.instance.termination;

import com.sequenceiq.cloudbreak.core.flow2.AbstractStackAction;
import com.sequenceiq.cloudbreak.core.flow2.restart.InitializeMDCContextRestartAction;
import com.sequenceiq.flow.core.FlowState;
import com.sequenceiq.flow.core.RestartAction;

enum InstanceTerminationState implements FlowState {

    INIT_STATE,
    TERMINATION_FAILED_STATE(InstanceTerminationFailureAction.class),
    TERMINATION_STATE(InstanceTerminationAction.class),
    TERMINATION_FINISHED_STATE(InstanceTerminationFinishedAction.class),
    FINAL_STATE;

    private Class<? extends AbstractStackAction<?, ?, ?, ?>> action;

    InstanceTerminationState() {
    }

    InstanceTerminationState(Class<? extends AbstractStackAction<?, ?, ?, ?>> action) {
        this.action = action;
    }

    @Override
    public Class<? extends AbstractStackAction<?, ?, ?, ?>> action() {
        return action;
    }

    @Override
    public Class<? extends RestartAction> restartAction() {
        return InitializeMDCContextRestartAction.class;
    }
}

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/instance/termination/InstanceTerminationState.kt
package com.sequenceiq.cloudbreak.core.flow2.stack.instance.termination

import com.sequenceiq.cloudbreak.core.flow2.AbstractAction
import com.sequenceiq.cloudbreak.core.flow2.FlowState

internal enum class InstanceTerminationState : FlowState {

    INIT_STATE,
    TERMINATION_FAILED_STATE(InstanceTerminationFailureAction::class.java),
    TERMINATION_STATE(InstanceTerminationAction::class.java),
    TERMINATION_FINISHED_STATE(InstanceTerminationFinishedAction::class.java),
    FINAL_STATE;

    private val action: Class<out AbstractAction<FlowState, FlowEvent, CommonContext, Payload>>

    private constructor() {
    }

    private constructor(action: Class<out AbstractAction<FlowState, FlowEvent, CommonContext, Payload>>) {
        this.action = action
    }


    override fun action(): Class<out AbstractAction<FlowState, FlowEvent, CommonContext, Payload>> {
        return action
    }
}
